engine/execution/ingestion/engine.go

```diff
@@ -930,344 +992,422 @@ func (e *Engine) generateExecutionReceipt(
-				}
+func (e *Engine) validateStateDelta(delta *messages.ExecutionStateDelta) error {
+	// must match the statecommitment for parent block
+	parentCommitment, err := e.execState.StateCommitmentByBlockID(e.unit.Ctx(), delta.ParentID())
+	if err != nil && !errors.Is(err, storage.ErrNotFound) {
+		return fmt.Errorf("could not get parent sttecommitment: %w", err)
+	}
 
-				if executableBlock.IsComplete() {
+	// the parent block has not been executed yet, skip
+	if errors.Is(err, storage.ErrNotFound) {
+		return nil
+	}
 
-					log.Debug().Msg("block complete - executing")
+	if !bytes.Equal(parentCommitment, delta.StartState) {
+		return engine.NewInvalidInputErrorf("internal inconsistency with delta for block (%v) - state commitment for parent retrieved from DB (%x) different from start state in delta! (%x)",
+			delta.ParentID(),
+			parentCommitment,
+			delta.StartState)
+	}
 
-					e.wg.Add(1)
-					go e.executeBlock(context.Background(), executableBlock)
-				}
-				e.syncInProgress.Store(false)
-				e.metrics.ExecutionSync(e.syncInProgress.Load())
-				log.Debug().Msg("final target sync block processed")
+	// TODO: validate the delta with the child block's statecommitment
 
-				return nil
-			})
+	return nil
+}
 
-		if err != nil {
-			log.Err(err).Msg("error while processing final target sync block")
-		}
+func (e *Engine) applyStateDelta(delta *messages.ExecutionStateDelta) {
+	blockID := delta.ID()
+	log := e.log.With().Hex("block", blockID[:]).Logger()
 
-		return
-	}
+	log.Debug().Msg("applying delta for block")
 
-	err = e.mempool.SyncQueues.Run(func(backdata *stdmap.QueuesBackdata) error {
+	// TODO - validate state delta, reject invalid messages
 
-		executionQueue, exists := backdata.ByID(executionStateDelta.Block.ID())
-		if !exists {
-			return fmt.Errorf("fatal error - synced delta not present in sync queue")
-		}
-		_, newQueues := executionQueue.Dismount()
-		for _, queue := range newQueues {
-			if !bytes.Equal(
-				queue.Head.Item.(*messages.ExecutionStateDelta).StartState,
-				executionReceipt.ExecutionResult.FinalStateCommit,
-			) {
-				return fmt.Errorf("internal incosistency with delta - state commitment for after applying delta different from start state of next one! ")
-			}
+	executionReceipt, err := e.saveExecutionResults(
+		e.unit.Ctx(),
+		&delta.ExecutableBlock,
+		delta.StateInteractions,
+		delta.Events,
+		delta.TransactionResults,
+		delta.StartState,
+	)
 
-			added := backdata.Add(queue)
-			if !added {
-				return fmt.Errorf("fatal error cannot add children block to sync queue")
-			}
+	if err != nil {
+		log.Fatal().Err(err).Msg("fatal error while processing sync message")
+	}
 
-			e.syncWg.Add(1)
-			go e.saveDelta(ctx, queue.Head.Item.(*messages.ExecutionStateDelta))
-		}
-		backdata.Rem(executionStateDelta.Block.ID())
-		return nil
-	})
+	if !bytes.Equal(executionReceipt.ExecutionResult.FinalStateCommit, delta.EndState) {
+		log.Fatal().
```

Comment:
Crashing a node is not a good idea here. We should return error to caller. Otherwise any malformed delta can introduce series of crashes